File: Core/src/main/java/com/github/ezauton/core/localization/estimators/EncoderRotationEstimator.java
package com.github.ezauton.core.localization.estimators;

import com.github.ezauton.core.localization.IRotationalLocationEstimator;
import com.github.ezauton.core.localization.ITranslationalLocationEstimator;
import com.github.ezauton.core.localization.Updateable;
import com.github.ezauton.core.trajectory.geometry.ImmutableVector;
import com.github.ezauton.core.utils.MathUtils;
import com.github.ezauton.core.localization.sensors.ITranslationalDistanceSensor;

/**
 * Describes an Updateable object that can track the location and heading of the robot using a rotational device
 * which can record angle (i.e. gyro) and a device which can record translational distance (i.e., encoder).
 */
public final class EncoderRotationEstimator implements IRotationalLocationEstimator, ITranslationalLocationEstimator, Updateable
{

    private final IRotationalLocationEstimator rotationalLocationEstimator;
    private final ITranslationalDistanceSensor distanceSensor;
    private double velocity;
    private double lastPosition;
    private double dPos;
    private ImmutableVector dPosVec;
    private ImmutableVector positionVec;
    private boolean init = false;

    /**
     * Create an EncoderRotationEstimator
     *
     * @param rotationalLocationEstimator An object that can estimate our current heading
     * @param distanceSensor              An encoder or encoder-like object.
     */
    public EncoderRotationEstimator(IRotationalLocationEstimator rotationalLocationEstimator, ITranslationalDistanceSensor distanceSensor)
    {
        this.rotationalLocationEstimator = rotationalLocationEstimator;
        this.distanceSensor = distanceSensor;
    }

    /**
     * Set the current position to <0, 0>, in effect resetting the location estimator
     */
    public void reset() //TODO: Reset heading
    {
        lastPosition = distanceSensor.getPosition();
        dPosVec = new ImmutableVector(0, 0);
        positionVec = new ImmutableVector(0, 0);
        init = true;
    }


    @Override
    public double estimateHeading()
    {
        return rotationalLocationEstimator.estimateHeading();
    }

    /**
     * @return The current velocity vector of the robot in 2D space.
     */
    @Override
    public ImmutableVector estimateAbsoluteVelocity()
    {
        return MathUtils.Geometry.getVector(velocity, rotationalLocationEstimator.estimateHeading());
    }


    /**
     * @return The current location as estimated from the encoders
     */
    @Override
    public ImmutableVector estimateLocation()
    {
        return positionVec;
    }

    /**
     * Update the calculation for the current heading and position. Call this as frequently as possible to ensure optimal results
     *
     * @return True
     */
    @Override
    public boolean update()
    {
        if(!init)
        {
            throw new IllegalArgumentException("Must be initialized! (call reset())");
        }
        if(rotationalLocationEstimator instanceof Updateable)
        {
            ((Updateable) rotationalLocationEstimator).update();
        }
        velocity = distanceSensor.getVelocity();
        dPos = distanceSensor.getPosition() - lastPosition;
        dPosVec = MathUtils.Geometry.getVector(dPos, rotationalLocationEstimator.estimateHeading());
        positionVec = positionVec.add(dPosVec);

        lastPosition = distanceSensor.getPosition();

        return true; //TODO: Return false sometimes?
    }
}

File: Core/src/main/kotlin/com/github/ezauton/core/localization/estimators/EncoderRotationEstimator.kt
package com.github.ezauton.core.localization.estimators

import com.github.ezauton.conversion.*
import com.github.ezauton.core.localization.RotLocEst
import com.github.ezauton.core.localization.TransLocEst
import com.github.ezauton.core.localization.Updatable
import com.github.ezauton.core.localization.sensors.TranslationalDistanceSensor
import com.github.ezauton.core.utils.math.polarVector2D

/**
 * Describes an object that can track the location and heading of the robot using a rotational device
 * which can record angle (i.e. gyro) and a device which can record translational distance (i.e., encoder).
 */
class EncoderRotationEstimator
/**
 * Create an EncoderRotationEstimator
 *
 * @param rotationalLocationEstimator An object that can estimate our current heading
 * @param distanceSensor An encoder or encoder-like object.
 */
  (
  private val rotationalLocationEstimator: RotLocEst,
  private val distanceSensor: TranslationalDistanceSensor
) : RotLocEst, TransLocEst, Updatable {
  private var velocity: LinearVelocity = zero()
  private var lastPosition: Distance  = zero()
  private var dPos: Distance = zero()
  private lateinit var dPosVec: ConcreteVector<Distance>
  private lateinit var positionVec: ConcreteVector<Distance>
  private var init = false

  /**
   * Set the current position to <0, 0>, in effect resetting the location estimator
   */
  fun reset() { // TODO: Reset heading
    lastPosition = distanceSensor.position
    dPosVec = vec(0.0, 0.0)
    positionVec = vec(0.0, 0.0)
    init = true
  }

  override fun estimateHeading(): Angle {
    return rotationalLocationEstimator.estimateHeading()
  }

  /**
   * @return The current velocity vector of the robot in 2D space.
   */
  override fun estimateAbsoluteVelocity(): ConcreteVector<LinearVelocity> {
    return polarVector2D(velocity, rotationalLocationEstimator.estimateHeading())
  }

  /**
   * @return The current location as estimated from the encoders
   */
  override fun estimateLocation() = positionVec

  /**
   * Update the calculation for the current heading and position. Call this as frequently as possible to ensure optimal results
   *
   * @return True
   */
  override fun update(): Boolean {
    if (!init) {
      throw IllegalArgumentException("Must be initialized! (call reset())")
    }
    if (rotationalLocationEstimator is Updatable) {
      (rotationalLocationEstimator as Updatable).update()
    }
    velocity = distanceSensor.velocity
    dPos = distanceSensor.position - lastPosition
    dPosVec = polarVector2D(dPos, rotationalLocationEstimator.estimateHeading())
    positionVec += dPosVec

    lastPosition = distanceSensor.position

    return true // TODO: Return false sometimes?
  }
}
